File: tests/eval/juex_eval/contract_oracle.py
from __future__ import annotations

import json
import pathlib
import re
from dataclasses import dataclass
from typing import Any


@dataclass
class ContractReport:
    passed: bool
    issues: list[str]

    def message(self) -> str:
        if not self.issues:
            return ""
        return ", ".join(self.issues)


def validate_agent_smoke_contract(conversation: pathlib.Path, events: pathlib.Path, token: str) -> ContractReport:
    issues: list[str] = []
    ok, detail = conversation_has_agent_smoke_tools(conversation, token)
    if not ok:
        issues.append(detail)
    ok, detail = events_have_agent_smoke_terminal_results(events, token)
    if not ok:
        issues.append(detail)
    return ContractReport(passed=not issues, issues=issues)


def conversation_has_agent_smoke_tools(path: pathlib.Path, token: str) -> tuple[bool, str]:
    if not path.is_file() and not path.is_dir():
        return False, "missing Thread Journal"
    tool_uses: dict[str, str] = {}
    seen_tools: set[str] = set()
    legacy_uses: list[str] = []
    saw_tty_exec = False
    saw_write_stdin = False
    saw_exec_result = False
    issues: list[str] = []
    messages = load_thread_journal_records(path, "message", issues)
    if issues:
        return False, issues[0]
    for line_number, message in enumerate(messages, start=1):
        blocks = message.get("blocks")
        if not isinstance(blocks, list):
            continue
        for block in blocks:
            if not isinstance(block, dict):
                continue
            block_type = block.get("type")
            if block_type == "tool_use":
                tool_name = block.get("tool_name")
                if isinstance(tool_name, str):
                    seen_tools.add(tool_name)
                    tool_use_id = str(block.get("tool_use_id") or "")
                    if tool_use_id:
                        tool_uses[tool_use_id] = tool_name
                input_value = block.get("input")
                if tool_name == "exec_command":
                    saw_tty_exec = saw_tty_exec or (isinstance(input_value, dict) and input_value.get("tty") is True)
                if tool_name == "write_stdin":
                    saw_write_stdin = True
                if tool_name in {"shell", "shell_input"}:
                    legacy_uses.append(f"{line_number}:{tool_name}")
            elif block_type == "tool_result":
                tool_use_id = str(block.get("tool_use_id") or "")
                content = str(block.get("content") or "")
                if tool_uses.get(tool_use_id) == "exec_command" and token in content and "Process exited with code 0" in content:
                    saw_exec_result = True
    if legacy_uses:
        return False, "Thread Journal contains unprojected shell tool_use: " + ", ".join(legacy_uses)
    required = {"read", "write", "edit", "grep", "exec_command", "write_stdin"}
    missing = sorted(required - seen_tools)
    if missing:
        return False, "missing required tool_use blocks: " + ", ".join(missing)
    if not saw_tty_exec:
        return False, "missing exec_command tool_use with tty:true"
    if not saw_write_stdin:
        return False, "missing write_stdin tool_use"
    if not saw_exec_result:
        return False, f"missing exec_command tool_result containing {token} and successful exit status"
    return True, ""


def events_have_agent_smoke_terminal_results(path: pathlib.Path, token: str) -> tuple[bool, str]:
    if not path.is_file() and not path.is_dir():
        return False, "missing Thread Journal"
    delta_count = 0
    saw_install = False
    saw_prompt = False
    saw_done = False
    saw_carriage_return = False
    saw_write_stdin_completed = False
    saw_structured_exec_running = False
    saw_structured_write_stdin_result = False
    issues: list[str] = []
    events = load_thread_journal_records(path, "event", issues)
    if issues:
        return False, issues[0]
    for event in events:
        payload = event.get("payload")
        if not isinstance(payload, dict):
            continue
        if event.get("type") == "tool.output_delta":
            delta_count += 1
        if event.get("type") == "tool.completed":
            result = payload.get("result")
            content = _terminal_content(payload)
            saw_install = saw_install or "INSTALL" in content
            saw_prompt = saw_prompt or "PROMPT approve install" in content
            saw_done = saw_done or f"TTY-DONE {token}" in content
            saw_carriage_return = saw_carriage_return or "\r" in content
            if payload.get("name") == "exec_command" and isinstance(result, dict):
                session_id = result.get("session_id")
                saw_structured_exec_running = saw_structured_exec_running or (
                    result.get("running") is True and _number_not_bool(session_id) and session_id > 0
                )
            if payload.get("name") == "write_stdin":
                saw_write_stdin_completed = True
                if isinstance(result, dict):
                    exit_code = result.get("exit_code")
                    saw_structured_write_stdin_result = saw_structured_write_stdin_result or (
                        result.get("running") is False
                        and _number_not_bool(exit_code)
                        and exit_code == 0
                        and f"TTY-DONE {token}" in content
                    )
    if delta_count != 0:
        return False, f"events persisted {delta_count} transient tool.output_delta records"
    missing = []
    if not saw_install:
        missing.append("INSTALL progress")
    if not saw_prompt:
        missing.append("interactive prompt")
    if not saw_done:
        missing.append("TTY-DONE token")
    if not saw_carriage_return:
        missing.append("carriage-return progress update")
    if not saw_write_stdin_completed:
        missing.append("write_stdin completion")
    if not saw_structured_exec_running:
        missing.append("structured exec_command running result")
    if not saw_structured_write_stdin_result:
        missing.append("structured write_stdin result")
    if missing:
        return False, "terminal events missing " + ", ".join(missing)
    return True, ""


def load_thread_journal_records(
    path: pathlib.Path,
    field: str,
    issues: list[str],
) -> list[dict[str, Any]]:
    """Read messages or events from registered Generation Journals.

    Direct JSONL records remain accepted for the deterministic eval harness,
    which builds in-memory oracle views rather than runtime persistence.
    """
    paths = thread_journal_paths(path)
    if not paths:
        issues.append(f"missing Thread Journal: {path}")
        return []
    records: list[dict[str, Any]] = []
    for journal in paths:
        try:
            lines = journal.read_text(encoding="utf-8", errors="replace").splitlines()
        except OSError as exc:
            issues.append(f"read Thread Journal: {exc}")
            continue
        for line_number, line in enumerate(lines, start=1):
            if not line.strip():
                continue
            try:
                value = json.loads(line)
            except json.JSONDecodeError as exc:
                issues.append(f"Thread Journal {journal.name} line {line_number} is invalid JSON: {exc}")
                continue
            if not isinstance(value, dict):
                issues.append(f"Thread Journal {journal.name} line {line_number} must be a JSON object")
                continue
            framed = value.get("data")
            if isinstance(framed, dict):
                value = framed
            facts = value.get("facts")
            if not isinstance(facts, list):
                records.append(value)
                continue
            for fact in facts:
                if not isinstance(fact, dict):
                    continue
                record = fact.get(field)
                if isinstance(record, dict):
                    records.append(record)
    return records


def thread_journal_paths(path: pathlib.Path) -> list[pathlib.Path]:
    if path.is_file():
        return [path]
    generation_dir = path / "generations"
    if not generation_dir.is_dir():
        return []
    metadata_path = path / "thread.json"
    if not metadata_path.is_file():
        return []
    try:
        metadata = json.loads(metadata_path.read_text(encoding="utf-8"))
        generations = metadata.get("generations")
        if not isinstance(generations, list):
            return []
        registered: list[pathlib.Path] = []
        for generation in generations:
            if not isinstance(generation, dict):
                return []
            generation_id = generation.get("generation_id")
            if (
                not isinstance(generation_id, str)
                or re.fullmatch(r"g[0-9]{6}", generation_id) is None
            ):
                return []
            registered.append(generation_dir / f"{generation_id}.jsonl")
        return registered
    except (OSError, json.JSONDecodeError):
        return []


def _terminal_content(payload: dict[str, Any]) -> str:
    content = payload.get("content")
    if isinstance(content, str) and content:
        return content
    outcome = payload.get("outcome")
    if not isinstance(outcome, dict):
        return content if isinstance(content, str) else ""
    block = outcome.get("block")
    if not isinstance(block, dict):
        return content if isinstance(content, str) else ""
    recorded = block.get("content")
    return recorded if isinstance(recorded, str) else ""


def _number_not_bool(value: Any) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool)

File: tests/eval/contract_oracle.go
package eval

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/juex-ai/juex/internal/llm"
)

type ContractArtifacts struct {
	ConversationPath string
	EventsPath       string
}

type ContractExpectations struct {
	RequiredTools         []string
	AllowLegacyShellTools bool
	RequireTTYExec        bool
	ExecResultToken       string
	Events                EventContractExpectations
}

type EventContractExpectations struct {
	MinOutputDeltas                   int
	RequireInstallProgress            bool
	RequireInteractivePrompt          bool
	DoneToken                         string
	RequireCarriageReturn             bool
	RequireWriteStdinCompleted        bool
	RequireStructuredExecRunning      bool
	RequireStructuredExecCompleted    bool
	RequireStructuredWriteStdinResult bool
}

type ContractReport struct {
	Passed bool            `json:"passed"`
	Issues []ContractIssue `json:"issues,omitempty"`
}

type ContractIssue struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

func (r ContractReport) Summary() string {
	if len(r.Issues) == 0 {
		return "contract passed"
	}
	parts := make([]string, 0, len(r.Issues))
	for _, issue := range r.Issues {
		parts = append(parts, issue.Message)
	}
	return strings.Join(parts, "; ")
}

func ValidateContractArtifacts(artifacts ContractArtifacts, expect ContractExpectations) ContractReport {
	var issues []ContractIssue
	conv := inspectConversationArtifact(artifacts.ConversationPath, expect)
	issues = append(issues, conv.issues...)
	eventSummary := inspectEventArtifact(artifacts.EventsPath, expect.Events.DoneToken)
	issues = append(issues, eventSummary.issues...)

	if len(expect.RequiredTools) > 0 {
		for _, tool := range expect.RequiredTools {
			if conv.toolNames[tool] == 0 {
				issues = append(issues, ContractIssue{Code: "conversation.tool.missing", Message: "missing required tool_use block: " + tool})
			}
		}
	}
	if !expect.AllowLegacyShellTools && len(conv.legacyShellUses) > 0 {
		issues = append(issues, ContractIssue{
			Code:    "conversation.shell.legacy",
			Message: "conversation contains legacy shell tool_use: " + strings.Join(conv.legacyShellUses, ", "),
		})
	}
	if expect.RequireTTYExec && !conv.sawTTYExec {
		issues = append(issues, ContractIssue{Code: "conversation.exec.tty", Message: "missing exec_command tool_use with tty:true"})
	}
	if expect.ExecResultToken != "" && !conv.sawExecResult {
		issues = append(issues, ContractIssue{
			Code:    "conversation.exec.result",
			Message: fmt.Sprintf("missing exec_command tool_result containing %s and successful exit status", expect.ExecResultToken),
		})
	}

	issues = append(issues, validateEventExpectations(eventSummary, expect.Events)...)
	return ContractReport{Passed: len(issues) == 0, Issues: issues}
}

type conversationContractSummary struct {
	issues          []ContractIssue
	toolNames       map[string]int
	toolUseNames    map[string]string
	legacyShellUses []string
	sawTTYExec      bool
	sawExecResult   bool
}

func inspectConversationArtifact(path string, expect ContractExpectations) conversationContractSummary {
	summary := conversationContractSummary{
		toolNames:    map[string]int{},
		toolUseNames: map[string]string{},
	}
	if path == "" {
		summary.issues = append(summary.issues, ContractIssue{Code: "conversation.path.missing", Message: "missing conversation path"})
		return summary
	}
	lines, err := readContractLines(path)
	if err != nil {
		summary.issues = append(summary.issues, ContractIssue{Code: "conversation.read", Message: "read conversation: " + err.Error()})
		return summary
	}
	for lineNumber, line := range lines {
		if strings.TrimSpace(line) == "" {
			continue
		}
		var msg llm.Message
		if err := json.Unmarshal([]byte(line), &msg); err != nil {
			summary.issues = append(summary.issues, ContractIssue{
				Code:    "conversation.json",
				Message: fmt.Sprintf("conversation line %d is invalid JSON: %v", lineNumber+1, err),
			})
			continue
		}
		for _, block := range msg.Blocks {
			switch block.Type {
			case llm.BlockToolUse:
				toolName := block.ToolName
				summary.toolNames[toolName]++
				if block.ToolUseID != "" {
					summary.toolUseNames[block.ToolUseID] = toolName
				}
				if toolName == "exec_command" && block.Input["tty"] == true {
					summary.sawTTYExec = true
				}
				if toolName == "shell" || toolName == "shell_input" {
					summary.legacyShellUses = append(summary.legacyShellUses, fmt.Sprintf("%d:%s", lineNumber+1, toolName))
				}
			case llm.BlockToolResult:
				if expect.ExecResultToken == "" {
					continue
				}
				if summary.toolUseNames[block.ToolUseID] == "exec_command" &&
					strings.Contains(block.Content, expect.ExecResultToken) &&
					strings.Contains(block.Content, "Process exited with code 0") {
					summary.sawExecResult = true
				}
			}
		}
	}
	return summary
}

type eventContractSummary struct {
	issues                        []ContractIssue
	outputDeltas                  int
	sawInstallProgress            bool
	sawInteractivePrompt          bool
	sawDone                       bool
	sawCarriageReturn             bool
	sawWriteStdinCompleted        bool
	sawStructuredExecRunning      bool
	sawStructuredExecCompleted    bool
	sawStructuredWriteStdinResult bool
}

func inspectEventArtifact(path string, doneToken string) eventContractSummary {
	var summary eventContractSummary
	if path == "" {
		summary.issues = append(summary.issues, ContractIssue{Code: "events.path.missing", Message: "missing events path"})
		return summary
	}
	lines, err := readContractLines(path)
	if err != nil {
		summary.issues = append(summary.issues, ContractIssue{Code: "events.read", Message: "read events: " + err.Error()})
		return summary
	}
	for lineNumber, line := range lines {
		if strings.TrimSpace(line) == "" {
			continue
		}
		var event map[string]any
		if err := json.Unmarshal([]byte(line), &event); err != nil {
			summary.issues = append(summary.issues, ContractIssue{
				Code:    "events.json",
				Message: fmt.Sprintf("events line %d is invalid JSON: %v", lineNumber+1, err),
			})
			continue
		}
		payload, _ := event["payload"].(map[string]any)
		if payload == nil {
			continue
		}
		switch event["type"] {
		case "tool.output_delta":
			text, _ := payload["text"].(string)
			summary.outputDeltas++
			summary.sawInstallProgress = summary.sawInstallProgress || strings.Contains(text, "INSTALL")
			summary.sawInteractivePrompt = summary.sawInteractivePrompt || strings.Contains(text, "PROMPT approve install")
			if doneToken != "" {
				summary.sawDone = summary.sawDone || strings.Contains(text, "TTY-DONE "+doneToken)
			} else {
				summary.sawDone = summary.sawDone || strings.Contains(text, "TTY-DONE")
			}
			summary.sawCarriageReturn = summary.sawCarriageReturn || strings.Contains(text, "\r")
		case "tool.completed":
			name, _ := payload["name"].(string)
			result, _ := payload["result"].(map[string]any)
			if name == "exec_command" && result != nil {
				sessionID := numberValue(result["session_id"])
				if result["running"] == true && sessionID > 0 {
					summary.sawStructuredExecRunning = true
				}
				if result["running"] == false && numberValue(result["exit_code"]) == 0 {
					summary.sawStructuredExecCompleted = true
				}
			}
			if name == "write_stdin" {
				summary.sawWriteStdinCompleted = true
				if result != nil &&
					result["running"] == false &&
					numberValue(result["exit_code"]) == 0 &&
					outputHasDoneToken(fmt.Sprint(result["output"]), doneToken) {
					summary.sawStructuredWriteStdinResult = true
				}
			}
		}
	}
	return summary
}

func outputHasDoneToken(output string, doneToken string) bool {
	if doneToken == "" {
		return strings.Contains(output, "TTY-DONE")
	}
	return strings.Contains(output, "TTY-DONE "+doneToken)
}

func validateEventExpectations(summary eventContractSummary, expect EventContractExpectations) []ContractIssue {
	var issues []ContractIssue
	if expect.MinOutputDeltas > 0 && summary.outputDeltas < expect.MinOutputDeltas {
		issues = append(issues, ContractIssue{
			Code:    "events.delta.count",
			Message: fmt.Sprintf("expected at least %d tool.output_delta events, saw %d", expect.MinOutputDeltas, summary.outputDeltas),
		})
	}
	if expect.RequireInstallProgress && !summary.sawInstallProgress {
		issues = append(issues, ContractIssue{Code: "events.delta.install", Message: "events missing INSTALL progress"})
	}
	if expect.RequireInteractivePrompt && !summary.sawInteractivePrompt {
		issues = append(issues, ContractIssue{Code: "events.delta.prompt", Message: "events missing interactive prompt"})
	}
	if expect.DoneToken != "" && !summary.sawDone {
		issues = append(issues, ContractIssue{Code: "events.delta.done", Message: "events missing TTY-DONE token"})
	}
	if expect.RequireCarriageReturn && !summary.sawCarriageReturn {
		issues = append(issues, ContractIssue{Code: "events.delta.carriage_return", Message: "events missing carriage-return progress update"})
	}
	if expect.RequireWriteStdinCompleted && !summary.sawWriteStdinCompleted {
		issues = append(issues, ContractIssue{Code: "events.write_stdin.completed", Message: "events missing write_stdin completion"})
	}
	if expect.RequireStructuredExecRunning && !summary.sawStructuredExecRunning {
		issues = append(issues, ContractIssue{Code: "events.exec.structured_running", Message: "events missing structured exec_command running result"})
	}
	if expect.RequireStructuredExecCompleted && !summary.sawStructuredExecCompleted {
		issues = append(issues, ContractIssue{Code: "events.exec.structured_completed", Message: "events missing structured exec_command completed result"})
	}
	if expect.RequireStructuredWriteStdinResult && !summary.sawStructuredWriteStdinResult {
		issues = append(issues, ContractIssue{Code: "events.write_stdin.structured", Message: "events missing structured write_stdin result"})
	}
	return issues
}

func readContractLines(path string) ([]string, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	var lines []string
	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return lines, nil
}

func numberValue(v any) float64 {
	switch n := v.(type) {
	case int:
		return float64(n)
	case int64:
		return float64(n)
	case float64:
		return n
	case json.Number:
		out, _ := n.Float64()
		return out
	default:
		return -1
	}
}
